elfloader-tool/src/arch-riscv/boot.c

```diff
@@ -19,7 +19,11 @@
 #define PT_LEVEL_2 2
 
 #define PT_LEVEL_1_BITS 30
+#if __riscv_xlen == 32
+#define PT_LEVEL_2_BITS 22
```

Comment:
Did this work in the past for RV32 just because we have been lucky the addresses in use were "choosen well"?

Reply:
Yes.  And because the elfloader and kernel elf fit inside a single page mapping. The subsequent PPN in the mapping code would have been wrong.